storage/provider/openstack.go

```diff
@@ -0,0 +1,328 @@
+		openstackCfgKeyUser,
+		openstackCfgKeySecrets,
+		openstackCfgKeyRegion,
+		openstackCfgKeyTenantName} {
+		if _, ok := cfg.Attrs()[cfgKey]; !ok {
+			return errors.NotAssignedf("requisite configuration was not set: %v", cfgKey)
+		}
+	}
+
+	return nil
+}
+
+type openstackVolumeSource struct {
+	openstackClient OpenstackAdapter
+}
+
+var _ storage.VolumeSource = (*openstackVolumeSource)(nil)
+
+// CreateVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) CreateVolumes(args []storage.VolumeParams) ([]storage.Volume, []storage.VolumeAttachment, error) {
+
+	vols := make([]storage.Volume, 0, len(args))
+	volAtt := make([]storage.VolumeAttachment, 0, len(args))
+
+	for _, a := range args {
+		if a.Provider != OpenstackProviderType {
+			logger.Debugf(
+				"openstack storage provider was passed a request for another provider type: %v",
+				a.Provider,
+			)
+			continue
+		}
+
+		if a.Attachment != nil {
+
+			createdVol, err := s.openstackClient.CreateVolume(a.Size, a.Tag.String())
+			if err != nil {
+				return nil, nil, err
+			}
+
+			// TODO(katco-): It would be better to check that the
+			// os-volume-manage extension is supported and pass that
+			// truth into this struct so that we don't make
+			// round-trips if we have to. Currently, goose's Nova
+			// client does not support this.
+			attachment, err := s.openstackClient.AttachVolume(
+				string(a.Attachment.InstanceId),
+				createdVol.VolumeId,
+				"", // TODO(katco): Fill this in when attachment support is added to parameters.
+			)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			volAtt = append(volAtt, attachment)
+
+		} else {
+
+			createdVol, err := s.openstackClient.CreateVolume(a.Size, a.Tag.String())
+			if err != nil {
+				return nil, nil, err
+			}
+
+			vols = append(vols, createdVol)
+		}
+	}
+
+	return vols, volAtt, nil
+}
+
+// DescribeVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) DescribeVolumes(volIds []string) ([]storage.Volume, error) {
+
+	// In most cases, it is quicker to get all volumes and loop
+	// locally than to make several round-trips to the provider.
+	cinderVols, err := s.openstackClient.GetVolumesSimple()
+	if err != nil {
+		return nil, err
+	}
+
+	// Map volume ID to volume info for faster lookup.
+	idToCinderVol := make(map[string]storage.Volume)
+	for _, volInfo := range cinderVols {
+		idToCinderVol[volInfo.VolumeId] = volInfo
+	}
+
+	volumes := make([]storage.Volume, len(volIds))
+	for volIdx, volId := range volIds {
+		cinderVol, ok := idToCinderVol[volId]
+		if !ok {
+			return nil, errors.Errorf("could not find volume: %s", volId)
+		}
+		volumes[volIdx] = cinderVol
+	}
+
+	return volumes, nil
+}
+
+// DestroyVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) DestroyVolumes(volIds []string) []error {
+
+	errors := make([]error, len(volIds))
+	for volIdx, volId := range volIds {
+		if err := s.openstackClient.DeleteVolume(volId); err != nil {
```

Comment:
Like Attach/Detach, DestroyVolumes needs to be idempotent and not fail if the volume doesn't exist.